qiskit_ibm_provider/ibm_backend.py

```diff
@@ -478,12 +478,14 @@ def run(
                 "ESP readout not supported on this device. Please make sure the flag "
                 "'use_measure_esp' is unset or set to 'False'."
             )
-        if not self.configuration().simulator:
-            self._deprecate_id_instruction(circuits)
 
         if isinstance(shots, float):
             shots = int(shots)
 
+        circuits_to_run = circuits
+        if not self.configuration().simulator:
+            circuits_to_run = self._deprecate_id_instruction(circuits)
```

Comment:
I'm not sure if the re-naming to `circuits_to_run` was for added code clarity, but just wanted to point out that `circuits = self._deprecate_id_instruction(circuits)` would also be okay at least for fixing this bug.